### PR TITLE
HWKAPM-600 Renamed ProducerInfo to SourceInfo to be more generic, and…

### PR DESCRIPTION
**Pull request by objectiser**
… also record other source node information and enable the information to be used to resolve the new CausedBy correlation scope for deriving communication details and therefore calculating completion time

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Hi Pavol, this is a large PR, but mainly due to the rename of ProducerInfo to SourceInfo. Could you do an initial review to see if you spot any issues? I need to do some integration tests to better test the trace completion time calculation for the different link types, but was thinking about doing that as a separate PR - let me know your thoughts?

**Comment by objectiser** [comment]
Reviewer notes:
* ProducerInfo and ProducerInfoCache renamed to be more generic SourceInfo and SourceInfoCache, to cater for 'CausedBy' correlation link which can reference back to any node type, not just Producer
* CommunicationDetails are now created for any correlation id found in an initial Consumer
* When building trace CompletionTime information, Communication objects are now created for all nodes, as they could potentially be the source for CausedBy links - however this can be optimised in the near future by using knowledge built up regarding which nodes are important

File: api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java [hunk 1]
```diff
@@ -27,18 +27,18 @@
 import org.hawkular.apm.api.utils.SerializationUtil;
 
 /**
- * This class represents information cached about a producer, to enable it to be
+ * This class represents information cached about a source node, to enable it to be
  * correlated to a consumer.
  *
  * @author gbrown
  */
-public class ProducerInfo implements Externalizable, ApmEvent {
```
Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java]: Why not implementing `ApmEvent`?

Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java]: This belongs to line 35

File: api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java [hunk 2]
```diff
@@ -70,31 +94,31 @@ public void setId(String id) {
     }
 
     /**
-     * @return the sourceUri
+     * @return the fragmentUri
```
Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java]: Did source uri represent source of the communication? What does represent fragmentUri now?

File: processors/tracecompletiontime-deriver/src/main/java/org/hawkular/apm/processor/tracecompletiontime/TraceCompletionInformationUtil.java [hunk 3]
```diff
@@ -45,13 +45,36 @@
      * @param fragmentBaseTime The base time for the fragment (ns)
      * @param n The node
      */
-    public static void initialiseLinks(TraceCompletionInformation ci, long fragmentBaseTime, Node n) {
+    public static void initialiseLinks(TraceCompletionInformation ci, long fragmentBaseTime, Node n,
+            StringBuffer nodeId) {
```
Comment (pavolloffay) [inline processors/tracecompletiontime-deriver/src/main/java/org/hawkular/apm/processor/tracecompletiontime/TraceCompletionInformationUtil.java]: Isn't `StringBuilder` more efficient here?

File: api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java (hunk 1 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java]: This was removed when I getId() was changed to return a list of ids - but later had to revert it. So I can add it back,

However it wasn't a problem as ProducerInfo was never used as an ApmEvent (which I think is used for storage) - but I'll add back anyway.

File: processors/tracecompletiontime-deriver/src/main/java/org/hawkular/apm/processor/tracecompletiontime/TraceCompletionInformationUtil.java (hunk 3 above)
Comment (objectiser) [inline processors/tracecompletiontime-deriver/src/main/java/org/hawkular/apm/processor/tracecompletiontime/TraceCompletionInformationUtil.java]: +1

File: api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java (hunk 2 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/model/events/SourceInfo.java]: It represented the URI used to trigger the fragment, so now is actually more accurate.

**Comment by pavolloffay** [comment]
review complete LGTM :+1:

**Comment by objectiser** [comment]
@pavolloffay Ok thanks Pavol, I'll fix the above points and once the integration tests are fixed I'll merge.

**Pull request merged**